rfd/0147-repo-fork-syncing.md

```diff
@@ -0,0 +1,60 @@
+ 
+#### Branch Protections
+ 
+The only branch that will be protected is the `sync/rebase` branch.  Other branches are expected to allow force pushes so they can be forcefully synced based off the upstream state.
+ 
+### High Level Implementation
+ 
+At a high level there are two primary components to this implementation:
+1. A GitHub action which will be scheduled as well as triggered on changes to the `sync/rebase` branch.  This action will perform the actual rebase action and commit the results to the respective branches above.
+2. A script that is invoked when the rebase can not be applied cleanly.  This can allow some custom rebase logic as a means to reduce how often the `sync/rebase` branch needs to be manually updated to avoid rebase conflicts.  A simple example is that workflow automations not desired in the repo copy may be removed in the `sync/rebase` branch.  That means that any changes to those committed workflows would result in a conflict.  However it can be safely resolved that any modified files removed in the `sync/rebase` branch can be resolved by simply removing.
+ 
+At a high level the action will do the following steps:
+1. Setup auth and any necessary repo specific configuration
+2. Checkout the latest upstream contents
+3. Fetch our custom changes
+4. Push the exact copy branches `sync/upstream-[UPSTREAM_BRANCH_NAME]` to our fork
+5. Rebase our `sync/rebase` branch on to the latest upstream changes, using the script for resolutions if necessary and possible
+6. Force push the rebased changes to the `main` branch on our repo copy
+ 
+### Teleport Security Scanning
+ 
+This tooling will be leveraged to address the gaps in our release branch scanning by doing the following:
+1. Create 3 repos (1 for each supported version): `teleport-sec_scan-1`, `teleport-sec_scan-2`, `teleport-sec_scan-3`.  `1` will reference our latest release branch, with each subsequent version inspecting a consecutively older version.
+2. Setup the sync automation as documented above
+3. Adjust Dependabot configuration to only notify for security updates on these branches
+4. Adjust CodeQL Configuration to scan on these branches rather than trying to capture all branch coverage within just `teleport`
+ 
+Step 4 is an optional addition, but since Dependabot results will be located here, it seems to co-locate other security reporting to these branches too.  This will also reduce the "Fixed" and "Reopened" flapping that can occur when changes are seen inconsistently between `master` and the release branches.
+ 
+With the automation determining what branch it needs to sync to based off the upstream branches, there is the concern that as soon as the next version is created all repos will be updated.  This will leave a small gap of time where the oldest (soon to be deprecated) release is not being scanned.  This time window seems short enough to be tolerable, but an alternative would be to create a fourth repo to ensure consistent coverage.
```

Comment:
Could the repos use a single source of truth for what the "current" version is?

Reply:
There is a single source of truth already being proposed.  That source of truth is the branches within the `teleport` repo.  The concern is that a new branch may be created before we want to rotate our scanning references.  I don't view this as a huge concern but wanted to explicitly call it out.

Reply:
I meant using a source less volatile than teleport branches, so we can explicitly dictate which versions we want scanned.

Reply:
Possible, any ideas what might be a good fit?  I was trying to avoid any additional manual steps, so if there is already a record of what versions are supported and released that I can pull, that would be a helpful alternative.

Reply:
A file in each of the `teleport-sec_scan-` repos would do it, or a file in the 1st repo that the others pull from if you don't want to write multiple PRs on every release. Or some other idea, this is mostly a gut reaction and not a well-thought suggestion.

Reply:
I don't think there is a need to make the target version explicit.  I have updated the RFD to use a 4 repo process to cover the small gap during major releases.  I would like to avoid anything that requires a manual step to update the scanning repo's target version.  Let me know if you have any concerns.